### PR TITLE
[incubator][vc] add resource syncer implementation

### PR DESCRIPTION
**Pull request by adohe**
Add resource syncer implementation. nit I am still working on pod syncer and some dependency issue. will finish this tomorrow

### CODE REVIEW

File: incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	case reconciler.DeleteEvent:
+		err := c.reconcilePodRemove(request.Cluster.Name, request.Namespace, request.Name, request.Obj.(*v1.Pod))
+		if err != nil {
+			return reconciler.Result{}, err
+		}
+	}
+	return reconciler.Result{}, nil
+}
+
+func (c *controller) reconcilePodUpdate(cluster, namespace, name string, pod *v1.Pod) error {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go]: Why pod creation is done in reconciling update event not the AddEvent?

File: incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+	pPod := newObj.(*v1.Pod)
+	conversion.MutatePod(targetNamespace, pPod)
+
+	innerCluster := c.multiClusterPodController.GetCluster(cluster)
+	client, err := clientset.NewForConfig(restclient.AddUserAgent(innerCluster.GetClientInfo().Config, "syncer"))
+	if err != nil {
+		return err
+	}
+	client.CoreV1().Pods(targetNamespace).Create(pod)
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go]: We should log err if any when updating super master and return the err.

File: incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	}
+	var client *clientset.Clientset
+	vPod := vPodObj.(*v1.Pod)
+	if vPod.Spec.NodeName != pod.Spec.NodeName {
+		innerCluster := c.multiClusterPodController.GetCluster(clusterName)
+		client, err = clientset.NewForConfig(restclient.AddUserAgent(innerCluster.GetClientInfo().Config, "syncer"))
+		if err != nil {
+			return
+		}
+		err = client.CoreV1().Pods(vPod.Namespace).Bind(&v1.Binding{
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go]: We should log err when binding fails.

File: incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+					}
+					utilruntime.HandleError(fmt.Errorf("unable to convert object %T to *v1.Pod in %T", obj, c))
+					return false
+				default:
+					utilruntime.HandleError(fmt.Errorf("unable to handle object in %T: %T", c, obj))
+					return false
+				}
+			},
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc: c.backPopulate,
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/controllers/pod/controller.go]: What if c.backPopulate fails? do we have retry here?

File: incubator/virtualcluster/cmd/syncer/app/options/options.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// Config return a syncer config object
+func (o *ResourceSyncerOptions) Config() (*syncerconfig.Config, error) {
+	c := &syncerconfig.Config{}
+
+	// Prepare kube clients
+	inClusterClient, leaderElectionClient, virtualClusterClient, superMasterClient, err := createClients(o.ClientConnection, o.SuperMaster, o.LeaderElection.RenewDeadline.Duration)
+	if err != nil {
+
```
Comment (zhuangqh) [inline incubator/virtualcluster/cmd/syncer/app/options/options.go]: why don't you return error here?

File: incubator/virtualcluster/pkg/syncer/controllers/configmap/controller.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		err := c.reconcileConfigMapRemove(request.Cluster.Name, request.Namespace, request.Name, request.Obj.(*v1.ConfigMap))
+		if err != nil {
+			return reconciler.Result{}, err
+		}
+	}
+	return reconciler.Result{}, nil
+}
+
+func (c *controller) reconcileConfigMapUpdate(cluster, namespace, name string, configMap *v1.ConfigMap) error {
+	targetNamespace := cluster + "-" + namespace;
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/controllers/configmap/controller.go]: take this conversion into the `conversion` package? this function would be referred around the vc project.
redundant `;`. please do `go fmt` before submit PR.

Comment (adohe) [reply incubator/virtualcluster/pkg/syncer/controllers/configmap/controller.go]: done.

File: incubator/virtualcluster/cmd/syncer/app/options/options.go (hunk 5 above)
Comment (adohe) [reply incubator/virtualcluster/cmd/syncer/app/options/options.go]: done.

**Pull request merged**